utils/compare/graph.py

```diff
@@ -0,0 +1,306 @@
+import pandas as pd
+from search import ElasticService
+import numpy as np
+
+def graph(oMetrics, nMetrics):
+    metrics = []
+    x=[]
+    y=[]
+    for index, row in oMetrics.iterrows():
+        test = "{}-{}".format(row['profile'], row['messageSize'])
+        value = "{}".format(row['throughput'])
+        x.append(value)
+        y.append(test)
+    old = {'y' : x,
+        'x' : y,
+        'name' : 'Previous results average',
+        'type' : 'bar',
+        'orientation' : 'v'}
+    x=[]
+    y=[]
+    for index, row in nMetrics.iterrows():
+        test = "{}-{}".format(row['profile'], row['messageSize'])
+        value = "{}".format(row['throughput'])
+        x.append(value)
+        y.append(test)
+    new = {'y' : x,
+        'x' : y,
+        'name' : 'Current results average',
+        'type' : 'bar',
+        'orientation' : 'v'}
+    metrics.append(old)
+    metrics.append(new)
+
+    return metrics
+
+
+def get_final_graph_metrics(metric_of_interest, divider,previous_data, current_data): 
+    x=[]
+    y=[]
+    for o, row in previous_data.iterrows():
+
+        test = metric_of_interest
+        value = row[metric_of_interest]
+        x.append(value/divider)
+        
+        y.append(test)
+    #print('previous x ' + str(x))
+    old = {'y' : x,
+        'x' : y,
+        'name' : 'Previous results ' + str(metric_of_interest),
+        'type' : 'bar',
+        'orientation' : 'v'}
+    x=[]
+    y=[]
+    for n, row in current_data.iterrows():
+        test = metric_of_interest
+        value = row[metric_of_interest]
+        x.append(value/divider)
+        y.append(test)
+    #print('final x ' + str(x))
+    new = {'y' : x,
+        'x' : y,
+        'name' : 'Current results ' + str(metric_of_interest),
+        'type' : 'bar',
+        'orientation' : 'v'}
+    metrics = []
+    metrics.append(old)
+    metrics.append(new)
+
+    return metrics
+
+def pd_merge(previous_data, current_data, columns):
+    pd_merged = current_data.merge(previous_data, how='outer', on=columns[:-1])
+    return pd_merged
+
+def find_result(row, tolerancy): 
+
+    if np.isinf(row['difference']): 
+        return "Pass"
+    # Fail if value is a more negative than tolerancy
+    # High positive valus mean new uuid is better
+    if row['difference'] < 0 and tolerancy > row['difference']: 
+        return 'Fail'
+    else: 
+        return "Pass"
+
+
+def tolerancy(all_data_points, tolerancy_percent): 
+    print('tolerancy ' + str(all_data_points))
+    all_data_points.pad(inplace=True)
+
+    result = all_data_points.iloc[:,[-2,-1]].pct_change(axis=1).iloc[:,1] * 100
+
+    all_with_result = all_data_points.assign(difference=result)
+
+    all_with_result['result'] = all_with_result.apply(lambda row: find_result(row, tolerancy_percent), axis=1)
+
+    return all_with_result
+
+def process_all_data(metric_of_interest, find_metrics,uuid, ids, data_func, index, divider, additional_columns): 
+    # Process all
+
+    columns = []
```

Comment:
Can we make it a `set` please, so that we can avoid duplicate columns.

Reply:
Because I use this variable as the columns for the dataframe I dont think it allows me to specify it as a set. 


  ```
File "/Users/prubenda/PycharmProjects/e2e-benchmarking/utils/compare/venv3/lib/python3.10/site-packages/pandas/core/frame.py", line 702, in __init__
    raise ValueError("columns cannot be a set")
ValueError: columns cannot be a set
```

Reply:
In that case, we can have set (to handle duplicates if any in the additionalColumns) to collect all the columns and then type cast it to a list which is supported by data frames.